arroyo-worker/src/engine.rs

```diff
@@ -130,11 +92,66 @@ pub trait StreamNode: Send {
     ) -> JoinHandle<()>;
 }
 
+pub struct WatermarkHolder {
+    // This is the last watermark with an actual value; this helps us keep track of the watermark we're at even
+    // if we're currently idle
+    last_present_watermark: Option<SystemTime>,
+    cur_watermark: Option<Watermark>,
```

Comment:
What about having Watermark cover the `None` case?

Reply:
In practice most of the handling is disjoint between the None and Some cases, and it makes using it much more convenient because we can use `?` for None.